encode_map.go

```diff
@@ -159,6 +214,41 @@ func encodeStructValue(e *Encoder, strct reflect.Value) error {
 	return nil
 }
 
+func encodeCanonicalStructValue(e *Encoder, strct reflect.Value, fields []*field) error {
+	// Hack to reuse the currently instantiated encoding with the
+	// provided options
+	realWriter := e.w
+	defer func() { e.w = realWriter }()
+
+	// Creates a slice which will be sorted before we write to the real writer.
+	values := make([][]byte, len(fields))
+
+	for i, f := range fields {
+		buf := &bytes.Buffer{}
+		e.w = buf
+		if err := e.EncodeString(f.name); err != nil {
+			return err
+		}
+		if err := f.EncodeValue(e, strct); err != nil {
+			return err
+		}
+		values[i] = buf.Bytes()
+	}
+
+	// We sort the serialized []byte representation of the (keys+values)
+	sortBytesArray(values)
```

Comment:
It is for simplicity and probably speed as well.

Sorting by key only or by (key+value) ends up being the same: It ensures canonical encoding.
The reason why using (key+value) is simpler is because using key only will add more code:
Right now I have: 
1/ one for loop to encode the entire map
2/ then I sort
3/ then I write the sorted result.

If I use only the keys, I would need to:
1/ Encode the keys (first loop), and create/build an new map where the key is a []byte (msgpack representation of the key) and the value is the value. 
2/ then I sort the msgpack encoded keys
3/ Then I write the sorted result. For each key I will need to retrieve the value in the map.

It means that in this second case, I'll have an extra map and an extra access to this temp map on the write.

Another option I could investigate would be to have an object with two fields {MsgPackKey, Value}, and sort this one based on the MsgPackKey and write. This way I could sort only by key.
MAybe it would be more efficient, I would need to run benchmarks